File: rust/examples/linux/other/mpr121/demo/src/main.rs
use linux_embedded_hal::{Delay, I2cdev};
use periph::chips::other::Mpr121Full;
use std::thread::sleep;
use std::time::Duration;

static NOTES: [&str; 12] = [
    "C4", "C#4", "D4", "D#4", "E4", "F4", "F#4", "G4", "G#4", "A4", "A#4", "B4",
];

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x5A);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus"); // Open I²C bus, (/dev/i2c-N) → I2cdev
    let mut delay = Delay;
    let mut chip = Mpr121Full::new(dev, addr, &mut delay).expect("init");         // Construct MPR121 Full, (i2c, addr=0x5A, delay) → Result<Mpr121Full, _>
    let mut previous: u16 = 0;

    loop {
        let mask = chip.touched().expect("touched");                                // Read 12-bit touch bitmask, () → Result<u16, _> bitmask
        let newly_pressed = mask & !previous;
        let newly_released = !mask & previous;
        for n in 0..12u8 {
            if newly_pressed & (1u16 << n) != 0 {
                println!("NOTE ON:  {}", NOTES[n as usize]);
            }
            if newly_released & (1u16 << n) != 0 {
                println!("NOTE OFF: {}", NOTES[n as usize]);
            }
        }
        previous = mask;
        sleep(Duration::from_millis(50));
    }
}

File: cpp/examples/linux/other/Mpr121/demo/main.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "Mpr121.h"

static const char* NOTES[12] = {
    "C4", "C#4", "D4", "D#4", "E4", "F4", "F#4", "G4", "G#4", "A4", "A#4", "B4"
};

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x5A;
    I2CConnectionLinux connection(bus, addr);                                 // Create I2C connection, (bus, addr) → I2CConnectionLinux

    MPR121Full mpr(connection);                                                 // Create MPR121 Full, (connection) → MPR121Full
    uint16_t previous = 0;

    while (true) {
        uint16_t mask = mpr.touched();                                          // Read 12-bit touch bitmask, () → uint16_t bitmask
        uint16_t newly_pressed = mask & ~previous;
        uint16_t newly_released = (~mask) & previous;
        for (uint8_t n = 0; n < 12; n++) {
            if (newly_pressed & (1u << n)) {
                printf("NOTE ON:  %s\n", NOTES[n]);
            }
            if (newly_released & (1u << n)) {
                printf("NOTE OFF: %s\n", NOTES[n]);
            }
        }
        previous = mask;
        usleep(50 * 1000);
    }
}
